first attempt to `stabilizer` for groups

ThomasBreuer opened the pull request:
This pull request deals with some natural actions of permutation and matrix groups.

ThomasBreuer committed:
first attempt to `stabilizer`

fingolfin commented on src/Groups/action.jl:
```diff
@@ -267,3 +267,43 @@ function on_indeterminates(f::Nemo.MPolyElem, s::PermGroupElem)
+
+function stabilizer(G::MatrixGroup{ET,MT}, pnt::AbstractAlgebra.Generic.FreeModuleElem{ET}) where {ET,MT}
+    img = GAP.GapObj(G.ring_iso.(pnt.v))[1]
+    return Oscar._as_subgroup(G, GAP.Globals.Stabilizer(G.X, img, GAP.Globals.OnRight))
```
Alternatively one could use the optional `gens, acts` arguments with the GAP generators and then the Oscar generators, then no ring isomorphism is needed, and one could even compute orbits this way.

ThomasBreuer replied:
Yes. Apparently there are no problems on the GAP side in the case of `stabilizer`. This means that we can get a generic method, see the version which I have just pushed,

fingolfin commented on src/Groups/action.jl:
```diff
@@ -267,3 +267,43 @@ function on_indeterminates(f::Nemo.MPolyElem, s::PermGroupElem)
+end
+
+function stabilizer(G::MatrixGroup{ET,MT}, pnt::Vector{AbstractAlgebra.Generic.FreeModuleElem{ET}}) where {ET,MT}
+    img = GAP.GapObj([GAP.GapObj(G.ring_iso.(x.v))[1] for x in pnt])
```
Shouldn't this be OnTuples? Or maybe it works because we are essentially doing matrix multiplication here? Alas, I wonder if that might prevent certain optimizations for OnTuples to kick in (namely one can compute a tuple stabilizer by iterating regular stabilizer computations over the tuple entries)

ThomasBreuer replied:
Yes, the code was correct because of the interpretation as matrix multiplication.
And yes, there is in fact one situation in the GAP library where an `OnTuples` stabilizer is computed iteratively.
(The code is meanwhile looking differently, this question does not arise anymore.)

fingolfin reviewed (approved):
Looks good to me. One could also do Tuples and more later on. Gotta think about how to best abstract this away

next: change the galois groups code to use this and remove the (set_)stabilizer methods in there?

ThomasBreuer committed:
next step for `stabilizer`

- changed `stabilizer` such that GAP uses the action on Julia objects via a Julia function; this way, we have a generic method plus several methods with a default action function (I was surprised that the GAP implementation did not cause problems)
- fixed `on_sets` for `Set`s (where we cannot assume that the elements can be sorted)
- added a `FreeModuleElem^MatrixGroupElem` method (needed because of default delegations, like in GAP)
- extended the tests
- extended the documentation (new Section on Group actions)

ThomasBreuer committed:
added a missing `hash` method

see Nemocas/AbstractAlgebra.jl/issues/1033

ThomasBreuer commented:
Just for curiosity: Why did the CI tests not run after my updates?

fingolfin commented:
To me it looks as if the CI tests did run after your last push (and succeeded)?

fingolfin commented on src/Groups/action.jl:
```diff
@@ -267,3 +269,60 @@ function on_indeterminates(f::Nemo.MPolyElem, s::PermGroupElem)
+stabilizer(G::MatrixGroup{ET,MT}, pnt::Set{AbstractAlgebra.Generic.FreeModuleElem{ET}}) where {ET,MT} = stabilizer(G, pnt, on_sets)
+
+#TODO: remove this as soon as AbstractAlgebra provides the function in question
+hash(v::AbstractAlgebra.Generic.FreeModuleElem, u::UInt) = hash(v.v, u)
```
See https://github.com/Nemocas/AbstractAlgebra.jl/issues/1033 and https://github.com/Nemocas/AbstractAlgebra.jl/issues/1034

fingolfin reviewed (approved):
Looks good to me!
